packages/edge-config/src/utils/tracing.ts

```diff
@@ -1,14 +1,82 @@
-import type { TracerProvider, Tracer } from '@opentelemetry/api';
+import type { TracerProvider, Tracer, Attributes } from '@opentelemetry/api';
 import { name as pkgName, version } from '../../package.json';
 
-let tracer: Tracer | undefined;
+let tracer: TracerProvider | undefined;
 
 export function setTracerProvider(tracerProvider: TracerProvider): void {
-  tracer = tracerProvider.getTracer(pkgName, version);
+  tracer = tracerProvider;
 }
 
-export function getTracer(): Tracer | undefined {
-  return tracer;
+function getTracer(): Tracer | undefined {
+  return tracer?.getTracer(pkgName, version);
+}
+
+export function trace<F extends (...args: any) => any>(
+  fn: F,
+  options: {
+    name: string;
+    tags?: Record<string, string | number | boolean>;
+    tagSuccess?: (
+      result: ReturnType<F> extends PromiseLike<infer U> ? U : ReturnType<F>,
+    ) => Attributes;
+    tagError?: (error: Error) => Attributes;
+  } = {
+    name: fn.name,
+  },
+): F {
+  const traced = function (this: unknown) {
+    const args = arguments as unknown as unknown[];
+    const that = this;
```

Comment:
What's `that` for? Can you just use `this` throughout?

Reply:
Yup, was concerned that it might not be consistent with the nested calls, but shouldn't be a problem.